packets1/disconnect_test.go

```diff
@@ -1,24 +1,78 @@
 package packets1
 
 import (
+	"bytes"
 	"reflect"
 	"testing"
 
 	"github.com/stretchr/testify/assert"
+
+	pkts "github.com/energomonitor/bisquitt/packets"
 )
 
-func TestDisconnectStruct(t *testing.T) {
+func TestDisconnectConstructor(t *testing.T) {
+	assert := assert.New(t)
+
 	duration := uint16(123)
 	pkt := NewDisconnect(duration)
 
-	if assert.NotNil(t, pkt, "New packet should not be nil") {
-		assert.Equal(t, "*packets1.Disconnect", reflect.TypeOf(pkt).String(), "Type should be Disconnect")
-		assert.Equal(t, duration, pkt.Duration, "Bad Duration value")
+	if pkt == nil {
+		t.Fatal("New packet should not be nil")
 	}
+
+	assert.Equal("*packets1.Disconnect", reflect.TypeOf(pkt).String(), "Type should be Disconnect")
+	assert.Equal(duration, pkt.Duration, "Bad Duration value")
 }
 
 func TestDisconnectMarshal(t *testing.T) {
 	pkt1 := NewDisconnect(75)
 	pkt2 := testPacketMarshal(t, pkt1)
 	assert.Equal(t, pkt1, pkt2.(*Disconnect))
 }
+
+func TestDisconnectUnmarshalInvalid(t *testing.T) {
+	assert := assert.New(t)
+
+	// Packet without Duration is correct.
+	buff := bytes.NewBuffer([]byte{
+		2,                     // Length
+		byte(pkts.DISCONNECT), // MsgType
+	})
+	_, err := ReadPacket(buff)
+	assert.Nil(err)
```

Comment:
I would put this test into a separate function and reserve this one only for invalid cases as the name suggests.

Alternatively, changing the name to something less specific like `TestDisconnectUnmarshal` would also do. But this approach would be inconsistent with the tests elsewhere.

Reply:
I am aware of this. I had two reasons to write it this way:

1. (as you have said) consistency with other packets tests
2. The correct and incorrect packets (as to their length) are entwined: Length 2 is correct, 3 is incorrect, 4 is correct, >=5 is incorrect.

I propose this solution: I will
1. rename the function to `TestDisconnectUnmarshal`
2. add a test for length=4. This case is already tested in `TestDisconnectMarshal` but it's probably better to repeat it here too to make things really explicit and readable (all possible packet lengths covered).

Is this ok with you?

Reply:
Fine with me.